iroh-net/src/magicsock/node_map.rs

```diff
@@ -84,22 +85,57 @@ enum NodeStateKey {
     IpPort(IpPort),
 }
 
-/// Source for a new node.
+/// The origin or *source* through which an address associated with a remote node
+/// was discovered.
 ///
-/// This is used for debugging purposes.
-#[derive(strum::Display, Debug)]
+/// An aggregate of the [`Source`]s of all the addresses of a node describe the
+/// [`Source`]s of the node itself.
+///
+/// A [`Source`] helps track how and where an address was learned. Multiple
+/// sources can be associated with a single address, if we have discovered this
+/// address through multiple means.
+///
+/// Each time a [`NodeAddr`] is added to the node map, usually through
+/// [`crate::endpoint::Endpoint::add_node_addr_with_source`], a [`Source`] must be supplied to indicate
+/// how the address was obtained.
+///
+/// A [`Source`] can describe a variety of places that an address or node was
+/// discovered, such as a configured discovery service, the network itself
+/// (if another node has reached out to us), or as a user supplied [`NodeAddr`].
+
+#[derive(Serialize, Deserialize, strum::Display, Debug, Clone, Eq, PartialEq, Hash)]
 #[strum(serialize_all = "kebab-case")]
-pub(crate) enum Source {
-    /// Node was loaded from the fs.
+pub enum Source {
+    /// Address was loaded from the fs.
     Saved,
-    /// Node communicated with us first via UDP.
+    /// A node communicated with us first via UDP.
     Udp,
-    /// Node communicated with us first via relay.
+    /// A node communicated with us first via relay.
     Relay,
-    /// Application layer added the node directly.
+    /// Application layer added the address directly.
     App,
+    /// The address was discovered by a discovery service.
+    #[strum(serialize = "{name}")]
+    Discovery {
+        /// The name of the discovery service that discovered the address.
+        name: String,
+    },
+    /// Application layer with a specific name added the node directly.
     #[strum(serialize = "{name}")]
-    NamedApp { name: &'static str },
+    NamedApp {
+        /// The name of the application that added the node
+        name: String,
+    },
+}
+
+impl Source {
+    /// Returns true if the [`Source::NamedApp`] is the same as the given string.
+    pub fn is_named(&self, named_app: &str) -> bool {
```

Comment:
nothing uses this api or am i missing something?  Also, this is kind of a really weird API...

I must admit I spent more than 5 minutes in the example of the local-swarm-discovery trying to figure out if you can write that `if let` statement nicer.

You can do slightly nicer with a match by putting the `if` condition on the match arm, but not much.  It was rather disappointing.  If you make the `name` field a `&'static str` you can do a pretty nice match and if-let, but with the giant caveat you can't use the constant for the service name anymore because that's not a literal.

Anyway, that's a lot of words just to say that I don't think this method is the answer.  Can we remove it?